library/L1_Peripheral/msp432p401r/system_controller.hpp

```diff
@@ -5,44 +5,421 @@
+    static constexpr bit::Mask kLowSpeedSubsystemClockDividerSelect =
+        bit::CreateMaskFromRange(28, 30);
+  };
+  /// Structure containing the necessary DCO configurations to generate a
+  /// desired frequency.
+  struct DcoConfiguration_t
+  {
+    /// The 3-bit select code to select the target DCO frequency range.
+    uint8_t frequency_select;
+    /// The 10-bit signed tuning value used to obtain the target DCO frequency.
+    int16_t tuning_value;
+  };
 
-  void SetPeripheralClockDivider(const PeripheralID &, uint8_t) const override
+  /// Reference to the structure containing the clock system control registers.
+  inline static CS_Type * clock_system = msp432p401r::CS;
+  /// Reference to the device descriptor tag-length-value (TLV) structure
+  /// containing the clock system calibration constants.
+  inline static TLV_Type * device_descriptors = msp432p401r::TLV;
+
+  /// Configures the digitally controlled clock (DCO) to produce the desired
+  /// target nominal frequency to drive the master clock.
+  ///
+  /// @param frequency The desired target frequency between 1 MHz and 48 MHz.
+  void SetSystemClockFrequency(
+      units::frequency::megahertz_t frequency) const override
   {
-  }
+    const DcoConfiguration_t kDcoConfiguration =
+        CalculateDcoConfiguration(frequency);
+
+    UnlockClockSystemRegisters();
+    {
+      clock_system->CTL0 = bit::Insert(clock_system->CTL0,
+                                       kDcoConfiguration.tuning_value,
+                                       Control0Register::kTuningSelect);
+      clock_system->CTL0 = bit::Insert(clock_system->CTL0,
+                                       kDcoConfiguration.frequency_select,
+                                       Control0Register::kFrequencySelect);
+      clock_system->CTL0 =
+          bit::Set(clock_system->CTL0, Control0Register::kEnable);
+    }
+    LockClockSystemRegisters();
+
+    WaitForClockReadyStatus(Clock::kMaster);
+    SetClockSource(Clock::kMaster, Oscillator::kDigitallyControlled);
+    SetClockSource(Clock::kSubsystemMaster, Oscillator::kDigitallyControlled);
 
+    speed_in_hertz = frequency;
+  }
+  /// Sets the divider for the subsystem master clock (HSMCLK).
+  ///
+  /// @param peripheral_divider Divider value to set. Only the following
+  ///                           dividers are available: 1, 2, 4, 8, 16, 32, 64,
+  ///                           128.
+  void SetPeripheralClockDivider(const PeripheralID &,
+                                 uint8_t peripheral_divider) const override
+  {
+    SetClockDivider(Clock::kSubsystemMaster, peripheral_divider);
+  }
+  /// @returns The current clock divider used for the subsystem master clock.
   uint32_t GetPeripheralClockDivider(const PeripheralID &) const override
   {
-    return 1;
+    constexpr uint8_t kDividerValues[] = { 1, 2, 4, 8, 16, 32, 64, 128 };
+    const uint8_t kDividerSelect       = static_cast<uint8_t>(
```

Comment:
```suggestion
    const uint8_t kDividerSelect       = static_cast<uint8_t>(
```
Why cast to uint8_t? Just keep it as uint32_t and use it as an index? Can the value within this register go beyond our look up table?